ppb_vector/vector2.py

```diff
@@ -1,97 +1,174 @@
+        return left
     else:
-        raise ValueError(f"Cannot use {value} as a vector-like")
+        return _find_lowest_type(left, right)
 
 
-class Vector2(Sequence):
+class Vector2:
+    x: float
+    y: float
+    _length: float
 
-    def __init__(self, x: Real, y: Real):
-        self.x = x
-        self.y = y
-        self.length = hypot(x, y)
+    def __init__(self, x: typing.SupportsFloat, y: typing.SupportsFloat):
+        try:
+            self.x = x.__float__()
+        except AttributeError:
+            raise TypeError(f"{type(x).__name__} object not convertable to float")
+        try:
+            self.y = y.__float__()
+        except AttributeError:
+            raise TypeError(f"{type(y).__name__} object not convertable to float")
 
     @classmethod
-    def convert(cls, value: VectorLike) -> 'Vector2':
+    def convert(cls: typing.Type[VectorOrSub], value: VectorLike) -> VectorOrSub:
         """
         Constructs a vector from a vector-like.
         """
-        return _mkvector(value, castto=type(cls))
+        # Use Vector2.convert() instead of type(self).convert() so that 
+        # _find_lowest_vector() can resolve things well.
+        if isinstance(value, cls):
+            return value
+        elif isinstance(value, Vector2):
+            return cls(value.x, value.y)
+        # FIXME: Allow all types of sequences
+        elif isinstance(value, (list, tuple)) and len(value) == 2:
+            return cls(value[0].__float__(), value[1].__float__())
+        # FIXME: Allow all types of mappings
+        elif isinstance(value, dict) and 'x' in value and 'y' in value and len(value) == 2:
+            return cls(value['x'].__float__(), value['y'].__float__())
+        else:
+            raise ValueError(f"Cannot use {value} as a vector-like")
+
+    @property
+    def length(self) -> float:
+        if not hasattr(self, '_length'):
+            self._length = hypot(self.x, self.y)
+        return self._length
 
-    def __len__(self) -> int:
+    def __len__(self: VectorOrSub) -> int:
         return 2
 
-    def __add__(self, other: VectorLike) -> 'Vector2':
+    def __add__(self: VectorOrSub, other: VectorLike) -> VectorOrSub:
+        rtype = _find_lowest_vector(type(other), type(self))
         try:
-            other = _mkvector(other)
+            other = Vector2.convert(other)
         except ValueError:
             return NotImplemented
-        rtype = type(other) if isinstance(other, Vector2) else type(self)
         return rtype(self.x + other.x, self.y + other.y)
 
-    def __sub__(self, other: VectorLike) -> 'Vector2':
+    def __sub__(self: VectorOrSub, other: VectorLike) -> VectorOrSub:
+        rtype = _find_lowest_vector(type(other), type(self))
         try:
-            other = _mkvector(other)
+            other = Vector2.convert(other)
         except ValueError:
             return NotImplemented
-        rtype = type(other) if isinstance(other, Vector2) else type(self)
```

Comment:
Is it a good idea to move the computation of `rtype` before the `try/except`?
It can be that `_find_lowest_vector` throws an exception in cases where we would previously have returned `NotImplemented`

Reply:
They don't throw exceptions, and they are internal functions meant for DRYing.